Add the control variates gradient estimator

rlouf opened the pull request:
We add the control variates gradient estimator for stochastic gradient MCMC algorithm. Control Variates require one gradient estimation on the whole dataset, which begs two questions that may be answered in subsequent PRs:

1. Shouldn't we compute `logposterior_center` in a separate `init` function, and thus propagate a `GradientState`?
2. How can we let users distribute this computation as they wish? This operation may need distributing with `pmap` and we should allow that.

This PR is part of an effort to port SGMCMCJAX to blackjax, see #289.

junpenglao commented on blackjax/sgmcmc/gradients.py:
```diff
@@ -53,9 +69,106 @@ def logposterior_estimator_fn(position: PyTree, data_batch: PyTree) -> float:
+
+    """
+    data_size = jax.tree_leaves(data)[0].shape[0]
+    logposterior_grad_estimator_fn = grad_estimator(
+        logprior_fn, loglikelihood_fn, data_size
+    ).estimate
```
```suggestion
   simple_grad_estimator = grad_estimator(
        logprior_fn, loglikelihood_fn, data_size
    ).estimate
    logposterior_grad_estimator_fn = lambda *x: simple_grad_estimator(*x)[0]
```

junpenglao replied:
Not very happy of this suggestion as well so feel free to resolve.

rlouf replied:
I agree that my code is not very satisfying too.

junpenglao commented:
High level question: here you also revert the change in https://github.com/blackjax-devs/blackjax/pull/293/ where you turn step size into a callable (so user can control how step size change during sampling), what is the reasoning behind the revert?

rlouf commented:
> High level question: here you also revert the change in https://github.com/blackjax-devs/blackjax/pull/293/ where you turn step size into a callable (so user can control how step size change during sampling), what is the reasoning behind the revert?

That's an important question. The current high-level interface of SgLD is:

```python
sgld = blackjax.sgld(grad_estimator, schedule)
```

Or for a constant schedule

```python
sgld = blackjax.sgld(grad_estimator, 1e-3)
```

And then to take a step

```python 
sgld.step(rng_key, state)
```

junpenglao commented:
So for low to mid level usage, user might be able to d osomething like:

```python
cosine_decay_scheduler = optax.cosine_decay_schedule(0.0001, decay_steps=total_steps, alpha=0.95)

for i in ...:  # could be in a jax.scan as well
  step_size = cosine_decay_scheduler(i)
  minibatch = ...
  gradients = grad_estimator(state, minibatch)
  state = sgld.step(rng_key, state, gradients, step_size)
```

rlouf commented on blackjax/sgmcmc/gradients.py:
```diff
@@ -52,10 +54,103 @@ def logposterior_estimator_fn(position: PyTree, data_batch: PyTree) -> float:
+                control_variate,
+                logposterior_grad_estimate,
+                grad_estimator_state.control_variate_grad,
+                logposterior_grad_center_estimate,
```
Thanks! I'd be curious to see your implementation, especially SVRG!

rlouf replied:
I took some time to think about the general API, and here is what I came up with for SVRG; everything else is a simplified version of this (no update for CV, no update and no init for the simple estimator). What do you think?

```python
import jax
import blackjax


svrg = blackjax.sgmcmc.gradients.svrg_estimator(logprior_fn, loglikelihood_fn)
sghmc = blackjax.sghmc(svrg)

grad_state = svrg.init(centering_position, data)

minibatch = next(data)
step_size = next(schedule)
position, grad_state, info = sghmc.step(  # uses svrg.grad internally
    rng_key,
    position,
    grad_state,
    minibatch,
    step_size
)

# Perform several SGMCMC steps
# And update the control variate
grad_state = svrg.update(grad_state, position, data)

_, rng_key = jax.random.split(rng_key)
minibatch = next(data)
step_size = next(schedule)
position, grad_state, info = sghmc.step(
    rng_key,
    position,
    grad_state,
    minibatch,
    step_size
)
```

junpenglao replied:
```python
# Perform several SGMCMC steps
# And update the control variate
grad_state = svrg.update(grad_state, position, data)
```
Is this going to be in a for loop or the num of steps are already encoded in the init of svrg?

rlouf commented:
@junpenglao @bstaber The behavior in a loop was not very clear indeed, so let me give a full example. I modified (hopefully improved) the behavior slightly. For the simple Robbins-Monro estimator we have:

```python
import jax
import blackjax
import blackax.sgmcmc.gradients as gradients # need an alias


schedule: Generator[float]
data = PyTree
batches: Generator[jax.numpy.DeviceArray]
position: PyTree

# Get the CV gradient estimator and SGHMC algorithm
grad_estimator = gradients.simple_estimator(logprior_fn, loglikelihood_fn, num_examples)
sghmc = blackjax.sgmcmc.sghmc()

rng_key = jax.random.PRNGKey(0)
for step in range(num_training_steps):
    _, rng_key = jax.random.split(rng_key)

    minibatch = next(batches)
    step_size = next(schedule)
    position, grad_state, info = sghmc.step(  # uses svrg.grad internally
        rng_key,
        position,
        grad_estimator,
        minibatch,
        step_size
    )
```

Now for the Control Variates estimator:

```python
import jax
import blackjax
import blackax.sgmcmc.gradients as gradients # need an alias


schedule: Generator[float]
data = PyTree
batches: Generator[jax.numpy.DeviceArray]
position: PyTree
centering_position: PyTree


# Get the CV gradient estimator and SGHMC algorithm
cv = gradients.cv(logprior_fn, loglikelihood_fn, num_examples)
sghmc = blackjax.sgmcmc.sghmc()

# Initialize the gradient state
# (SGHMC state is simply the position)
grad_estimator = cv.init(centering_position, data)

rng_key = jax.random.PRNGKey(0)
for step in range(num_training_steps):
    _, rng_key = jax.random.split(rng_key)

    minibatch = next(batches)
    step_size = next(schedule)
    position, grad_state, info = sghmc.step(
        rng_key,
        position,
        grad_estimator,
        minibatch,
        step_size
    )
```

SVRG is a CV estimator with updates. @bstaber's intuition is correct, and we can re-use the same code as for CV; we just need to re-initialize the control variate every `cv_update_rate` steps:

```python
import jax
import blackjax


schedule: Generator[float]
data = PyTree
batches: Generator[jax.numpy.DeviceArray]
position: PyTree
centering_position: PyTree
cv_update_rate: int


# Get the CV gradient estimator and SGHMC algorithm
svrg = gradients.cv(logprior_fn, loglikelihood_fn, num_examples)
sghmc = blackjax.sghmc(cv)

# Initialize the gradient state
# (SGHMC state is simply the position)
grad_estimator = svrg.init(centering_position, data)

rng_key = jax.random.PRNGKey(0)
for step in range(num_training_steps):
    _, rng_key = jax.random.split(rng_key)

    minibatch = next(batches)
    step_size = next(schedule)
    position, grad_state, info = sghmc.step(
        rng_key,
        position,
        grad_estimator,
        minibatch,
        step_size
    )

    # SVRG is nothing more than CV that you can update
    if step % == cv_update_rate:
        grad_estimator = svrg.init(centering_position, data)
```

While it is naively tempting to compute the gradient estimate outside of `sghmc.step`, SGHMC (and some other algorithms) needs to compute the gradient several times before returning a sample; the situation is not quite the same as Optax.

We may go a step further to remove the awkwardness that @junpenglao saw in the code, and make CV effectively a wrapper around the Robbins-Monto estimator:

```python
grad_estimator = gradients.simple_estimator(logprior_fn, loglikelihood_fn, num_examples)
cv_grad_estimator = gradients.cv(grad_estimator, centering_position, data)
```

rlouf commented:
@junpenglao This is ready for review. I implemented the control variates as a wrapper around the simple estimator, as dicussed above. The code is much cleaner than before, and svrg is obtained by calling `gradients.control_variates` again on the gradient estimator within the sampling loop.

rlouf commented:
@junpenglao ping
